frontend/src/routes/Governance/common/util.tsx

```diff
@@ -650,18 +654,97 @@ export function hasInformOnlyPolicies(items: Array<PolicyTableItem>): boolean {
   return true
 }
 
+// Transform string annotation to key and value pair type
+/* istanbul ignore next */
+export const parseStringMap = (
+  anoString: string
+): {
+  [key: string]: string
+} => {
+  const parsed = anoString?.split('; ') ?? []
+  const result: {
+    [key: string]: string
+  } = {}
+  for (const keyEqualVal of parsed) {
+    const kv = keyEqualVal.split('=')
+    if (kv.length == 2) {
+      result[kv[0]] = kv[1]
+    }
+  }
+
+  return result
+}
+
+// type guard
+function isPolicy(object: Policy | DiscoveredPolicyItem): object is Policy {
+  return object && (object as Policy).metadata !== undefined
+}
+
 export function getPolicySource(
-  policy: Policy,
+  policy: Policy | DiscoveredPolicyItem,
   helmReleases: HelmRelease[],
   channels: Channel[],
   subscriptions: Subscription[],
   t: TFunction<string, undefined>
 ): string | JSX.Element {
-  const isExternal = resolveExternalStatus(policy)
+  if (!policy) {
+    // Prevent error in detail page
+    return <></>
+  }
+
+  const isExternal = policy && isPolicy(policy) ? resolveExternalStatus(policy) : policy._isExternal || false
   let source: string | JSX.Element = t('Local')
   if (isExternal) {
-    const policySource = resolveSource(policy, helmReleases, channels, subscriptions)
+    const policySource = isPolicy(policy)
+      ? resolveSource(policy.metadata.annotations ?? {}, helmReleases, channels, subscriptions)
+      : resolveSource(parseStringMap(policy.annotation), helmReleases, channels, subscriptions)
     source = policySource ? getSource(policySource, isExternal, t) : t('Managed externally')
   }
   return source
 }
+
+export function getEngineString(kind: string): string {
+  switch (kind) {
+    case 'ConfigurationPolicy':
+    case 'CertificatePolicy':
+    case 'OperatorPolicy':
+      return 'Open Cluster Management'
+    default:
+      return 'Unknown'
+  }
+}
+
+export function getEngineWithSvg(kind: string): JSX.Element {
+  switch (kind) {
+    case 'ConfigurationPolicy':
+    case 'CertificatePolicy':
+    case 'OperatorPolicy':
+      return (
+        <div style={{ display: 'flex', alignItems: 'center', gap: 8 }}>
+          <div
+            style={{
+              height: 18,
+              width: 18,
+            }}
+          >
+            {' '}
+            <OcmSvg />
+          </div>{' '}
+          Open Cluster Management
+        </div>
+      )
+    default:
+      return <>Unknown</>
+  }
+}
+
+/* istanbul ignore next */
+export const parseDiscoveryPolicies = (data: any): any => {
+  return JSON.parse(JSON.stringify(data), (k, v: string) => {
```

Comment:
Is it necessary to JSON stringify and parse it? Could you just have if statements to see if `disabled`, `_isExternal`, and/or `_hubClusterResource`   are set and directly modify them?

Reply:
I use this for both array and object. When I use this for an array it is used in worker. So we don't need to worry about memory issues. The other case is in unit tests.